### ISSUE TITLE
Dangling Federations and Federation Structure

### ISSUE THREAD

**Issue by moltzaum** [apache/trafficcontrol#2592]
`POST /cdns/foo/federations/`
Does not connect a deliveryservice, but it should. The method to connect deliveryservies is via
`POST federations/:id/deliveryservices`
However, the only way to get the required id is by checking the response from the first post. Logically, this should be one step, especially due to the visibility limitation on federations. If a post to `cdns/foo/federations` does not save the id (due to user error), it is impossible to see or delete the dangling federation that was just created.

The Visibility Limitation on Federations:
`GET /cdns/foo/federations`
     limits the scope of a search by cdn. If no ds is linked, no cdn is associated.
`GET /cdns/foo/federations/:id`
     currently ignores the scope of foo, but _theoretically_ tenacy via the ds should also limit the scope meaning a federation with no linked ds will not show up. Tenacy is relatively new, so I don't think this is currently the case. However, it is clear either way that it doesn't make sense to verify the id by a GET request like this.

I was initially going to try creating an optional deliveryservice field in the json request (matching the same structure of the GET response below). I was thinking that the optional field could be made required in a new major version, but I'm not certain it is the best solution due to the awkward structure. 

```
{
    "ttl": 48,
    "id": 3,
    "description": "the description",
    "cname": "new.cname.",
    "deliveryService": {
        "xmlId": "ds1",
        "id": 81
    }
}
```
Ideally, "deliveryservice" would be a list of ids in a similar fashion to the` federations/:id/deliveryservices` endpoint, but I'm hesitant to create a different structure between POST and GET responses. (contrary to the json above, federations and deliveryservices are a many to many relationship)

Back to my primary point, the overall goal I had in exploring this topic is ensuring a federation is always visible from some perspective. A new GET endpoint for all detached/unassociated federations is possible, but didn't seem to make the most sense to me. A different approach that I think is perhaps more valid is having stronger connections between the cdn name specified in the url and the federation. For instance: 
`POSTing to /cdns/cdn1/federations`,
then linking to ds2 (corresponding to cdn2) has absolutely NO relation to cdn1, other than there is a related tenant (except in the current implementation, cdn1 is completely ignored by the handler).

If posting to cdn1 actually associated a federation with the ds directly associated with cdn1 I think that would make much more sense, however it would change behavior. What is the likelyhood that something like this can be implemented?

**Comment by mitchell852** [comment]
I don't know the best answer here. IMO, this is the result of poor API design when federations were originally implemented. 

For example, what would you expect GET /api/federations to return? I would expect it to return the contents of the federation table but that is not the case. From what I can tell it returns "federation resolvers for federations assigned to the current user" or something. 

https://github.com/apache/trafficcontrol/blob/master/traffic_ops/app/lib/API/Federation.pm#L178

When I added the federation functionality to TP, i needed the following functionality in the API:

1. get federations
2. create a federation
3. update/delete a federation
4. assign a federation to delivery service(s)
5. assign a federation to user(s)
6. create a federation resolver
7. assign federation resolvers to a federation

so i wanted to create the following api endpoints:

1. GET /api/1.3/federations
2. POST /api/1.3/federations
3. PUT/DELETE /api/1.3/federations/:id
4. POST /api/1.3/federations/11/deliveryservices
5. POST /api/1.3/federations/11/users
6. POST /api/1.3/federation_resolvers
7. POST /api/1.3/federations/11/federation_resolvers

but unfortunately 1 and 2 were already taken so i had to get creative so i settled on:

1. GET /api/1.3/cdns/:name/federations
2. POST /api/1.3/cdns/:name/federations

and for consistency i decided to make 

3. PUT/DELETE /api/1.3/cdns/:name/federations/:fedId

because in my brain at that time i figured all federations were linked to a cdn via the ds linked to the federation but in hindsight i see the flaw in that thinking because as you've pointed out if you fail to link a federation to a ds, then you have no way of retrieving the "dangling" federation because neither of these will work:

- GET /api/1.3/federations <-- doesn't even return federations
- GET /api/1.3/cdns/:name/federations

So what to do? I like your idea of adding a new, optional field to the request json provided to POST /api/1.3/cdns/:name/federations. The field is something like:

dsIds: [ 1, 2, 3 ]

and make sure it's added to the documentation. unfortunately we can't make that field required without revving the api version and i'm not exactly sure what kind of effort that would entail

**Comment by moltzaum** [comment]
Suggested POST ("deliveryServices" is new)
```
{
    "ttl": 48,
    "id": 3,
    "description": "the description",
    "cname": "new.cname.",
    "deliveryServices": [ 1, 2, 3 ]
}
```

Current GET response
```
{
    "ttl": 48,
    "id": 3,
    "description": "the description",
    "cname": "new.cname.",
    "deliveryService": {
        "xmlId": "ds1",
        "id": 81
    }
}
```

As we discussed before, I think we both agree that an array of dsIds would be ideal for the json structure. However, do you think it is a good idea to include an array on one end, and a single entry on the other?

**Comment by mitchell852** [comment]
no, i don't particularly like them mismatching. i'd rather the GET response looks like

`{
    "ttl": 48,
    "id": 3,
    "description": "the description",
    "cname": "new.cname.",
    "deliveryServices": [ 1, 2, 3 ]
}`

but that would technically be breaking the api so i suppose just leave it
